services/AdReferralService.ts

```diff
@@ -0,0 +1,163 @@
+// Information about the campaign that referred the user to this page.
+export const KEY = 'session-referring-campaign';
+
+const MAX_SESSION_MS = 1000 * 60 * 60 * 24 * 30; // 30 days
+
+/**
+ * AdReferralService persists UTM parametes in Local Storage and allows for
+ * later retrieval of those parameters up until the data expires.
+ */
+class AdReferralService {
```

Comment:
🌈 this whole thing feels kinda over-engineered and not react-y to me, but i won't block on it. (also maybe that opinion is just due to opinions of the people i normally work with. 😆)

i prob would have just made a hook like this (super quick untested napkin code):

```ts
// zod (already installed in this app) makes parsing json at runtime super easy and 100% typ-esafe
const ReferralParams = z.object({
	utm_source: z.string().nullable().default(null),
	utm_medium: z.string().nullable().default(null),
	utm_campaign: z.string().nullable().default(null),
	utm_content: z.string().nullable().default(null),
})

const ReferralParamsStorage = z.object({
    data: ReferralParams,
	createdAt: z.number(),
})

const useReferralParams = () => {
	// next.js has a useRouter hook so you don't need to be directly interacting w/ the window
	const router = useRouter()
    // reads type-safe query params from url
	const queryParams = useMemo(() => {
		const parsed = ReferralParams.safeParse(router.query)
		if (!parsed.success || !some(Object.values(parsed.data))) {
			return null
		}
		return parsed.data
	}, [router.query])

	// i recommend useLocalStorage from https://github.com/juliencrn/usehooks-ts
	const [getLocalStorage, setLocalStorage, clearLocalStorage] = useLocalStorage(KEY)

    // persists query params to localstorage
	useEffect(() => {
		if (!queryParams) {
			return
		}
		setLocalStorage({ data: queryParams, createdAt: Date.now() })
	}, [queryParams])

   // reads query params back out from localstorage & parses & makes sure they aren't too old
	const getValue = useCallback(() => {
		const val = getLocalStorage()
		if (!val) {
			return null
		}
		const parsed = ReferralParamsStorage.safeParse(val)
		if (!parsed.success) {
			return null
		}
		const {createdAt, data} = parsed.data
		if (new Date(createdAt) < sub(neq Date(), {days: 30})) { //`sub` is from date-fns (already installed)
			return null
		}
		return data
	}, [getLocalStorage])

	return getValue
}
```